### ISSUE TITLE
Data race in Marshal() when the protobuf has extensions

### ISSUE THREAD

**Issue by stomakun** [gogo/protobuf#511]
**What version of protobuf and what language are you using?**
Version: 1ef32a8b9fc3f8ec940126907cedb5998f6318e4

**What did you do?**
If possible, provide a recipe for reproducing the error.
A complete runnable program is good with `.proto` and `.go` source code.

Create a message that has extensions. Generate `.pb.go`.
Create a struct of that message in Go.
Call Marshal() on that struct in two goroutines with go race detector enabled.

**What did you expect to see?**

Two identical byte slices.

**What did you see instead?**

```
==================
WARNING: DATA RACE
Read at 0x00c42084c800 by goroutine 122:
  github.com/gogo/protobuf/proto.(*XXX_InternalExtensions).extensionsWrite()
      go/src/github.com/gogo/protobuf/proto/extensions.go:133 +0x3c
  <something>.github.com/gogo/protobuf/proto.extensionsWrite()
      <autogenerated>:1 +0x49
  github.com/gogo/protobuf/proto.SizeOfInternalExtension()
      go/src/github.com/gogo/protobuf/proto/extensions_gogo.go:67 +0x3e
  <something>.ProtoSize()
      <something>.pb.go:?? +0x2ef
  <something>.Marshal()
      <something>.pb.go:?? +0x3c
  google.golang.org/grpc/encoding/proto.codec.Marshal()
      go/src/google.golang.org/grpc/encoding/proto/proto.go:70 +0x22c
  google.golang.org/grpc/encoding/proto.(*codec).Marshal()
      <autogenerated>:1 +0x64
  google.golang.org/grpc.encode()
      go/src/google.golang.org/grpc/rpc_util.go:487 +0x7c
  ...

Previous write at 0x00c42084c800 by goroutine 121:
  github.com/gogo/protobuf/proto.(*XXX_InternalExtensions).extensionsWrite()
      go/src/github.com/gogo/protobuf/proto/extensions.go:134 +0xb9
  <something>.github.com/gogo/protobuf/proto.extensionsWrite()
      <autogenerated>:1 +0x49
  github.com/gogo/protobuf/proto.SizeOfInternalExtension()
      go/src/github.com/gogo/protobuf/proto/extensions_gogo.go:67 +0x3e
  <something>.ProtoSize()
      <something>.pb.go:?? +0x2ef
  <something>.Marshal()
      <something>.pb.go:?? +0x3c
  google.golang.org/grpc/encoding/proto.codec.Marshal()
      go/src/google.golang.org/grpc/encoding/proto/proto.go:70 +0x22c
  google.golang.org/grpc/encoding/proto.(*codec).Marshal()
      <autogenerated>:1 +0x64
  google.golang.org/grpc.encode()
      go/src/google.golang.org/grpc/rpc_util.go:487 +0x7c
  ...
```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment?**

Basically, one would expect it safe to run multiple `Marshal()`s in parallel, since they conceptually only read the struct.
However, because of https://github.com/gogo/protobuf/blob/master/proto/extensions_gogo.go#L154 (which is called in `Marshal()`
in the generated codes), this is not true.

**Comment by donaldgraham** [comment]
Great job finding and reporting this.  Would you like to attempt a fix?
